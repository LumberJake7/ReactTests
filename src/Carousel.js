import { useState } from "react";
import "./Carousel.css";
import Card from "./Card";

/** Carousel: displays images and arrows to navigate through them
 *
 * Props:
 * - photos: array of {src, caption} objects
 * - title: string describing the collection of images
 *
 * State:
 * - currCardIdx: integer for current card index
 *
 * App --> Carousel --> Card
 */
function Carousel({ photos, title }) {
  const [currCardIdx, setCurrCardIdx] = useState(0);

  const currCard = photos[currCardIdx];
  const total = photos.length;

  function goForward() {
    if (currCardIdx < total - 1) {
      setCurrCardIdx(currCardIdx + 1);
    }
  }

  function goBackwards() {
    if (currCardIdx > 0) {
      setCurrCardIdx(currCardIdx - 1);
    }
  }

  return (
    <div className="Carousel">
      <h1>{title}</h1>
      <div className="Carousel-main">
        {currCardIdx > 0 && (
          <i
            className="bi bi-arrow-left-circle"
            role="button"
            aria-label="left arrow"
            onClick={goBackwards}
          />
        )}
        <Card
          caption={currCard.caption}
          src={currCard.src}
          currNum={currCardIdx + 1}
          totalNum={total}
        />
        {currCardIdx < total - 1 && (
          <i
            className="bi bi-arrow-right-circle"
            role="button"
            aria-label="right arrow"
            onClick={goForward}
          />
        )}
      </div>
    </div>
  );
}

export default Carousel;
