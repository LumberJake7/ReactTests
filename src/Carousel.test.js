// Carousel.test.js
import React from "react";
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Carousel from "./Carousel";

const TEST_IMAGES = [
  { src: "image1.jpg", caption: "Caption 1" },
  { src: "image2.jpg", caption: "Caption 2" },
  { src: "image3.jpg", caption: "Caption 3" },
];

// Smoke test
it("renders without crashing", () => {
  render(<Carousel photos={TEST_IMAGES} title="Test Carousel" />);
});

// Snapshot test
it("matches snapshot", () => {
  const { asFragment } = render(
    <Carousel photos={TEST_IMAGES} title="Test Carousel" />
  );
  expect(asFragment()).toMatchSnapshot();
});

// Test for goBackwards function
it("goes to the previous image", () => {
  const { container, getByRole } = render(
    <Carousel photos={TEST_IMAGES} title="Test Carousel" />
  );

  const rightArrow = getByRole("button", { name: /right arrow/i });
  fireEvent.click(rightArrow);

  expect(container.querySelector('img[alt="Caption 2"]')).toBeInTheDocument();

  const leftArrow = getByRole("button", { name: /left arrow/i });
  fireEvent.click(leftArrow);

  expect(container.querySelector('img[alt="Caption 1"]')).toBeInTheDocument();
});

// Test for goForward function
it("goes to the next image", () => {
  const { container, getByRole } = render(
    <Carousel photos={TEST_IMAGES} title="Test Carousel" />
  );

  expect(container.querySelector('img[alt="Caption 1"]')).toBeInTheDocument();

  const rightArrow = getByRole("button", { name: /right arrow/i });
  fireEvent.click(rightArrow);

  expect(container.querySelector('img[alt="Caption 2"]')).toBeInTheDocument();
});

// Test to ensure left arrow is hidden on the first image
it("hides the left arrow on the first image", () => {
  const { queryByRole } = render(
    <Carousel photos={TEST_IMAGES} title="Test Carousel" />
  );

  expect(
    queryByRole("button", { name: /left arrow/i })
  ).not.toBeInTheDocument();

  expect(queryByRole("button", { name: /right arrow/i })).toBeInTheDocument();
});

// Test to ensure right arrow is hidden on the last image
it("hides the right arrow on the last image", () => {
  const { getByRole, queryByRole } = render(
    <Carousel photos={TEST_IMAGES} title="Test Carousel" />
  );

  const rightArrow = getByRole("button", { name: /right arrow/i });
  fireEvent.click(rightArrow);
  fireEvent.click(rightArrow);

  expect(
    queryByRole("button", { name: /right arrow/i })
  ).not.toBeInTheDocument();

  expect(getByRole("button", { name: /left arrow/i })).toBeInTheDocument();
});
