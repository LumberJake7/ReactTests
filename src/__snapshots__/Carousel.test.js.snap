// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Carousel"
  >
    <h1>
      Test Carousel
    </h1>
    <div
      class="Carousel-main"
    >
      <div
        class="Card"
      >
        <h4
          class="Card-title"
        >
          Caption 1
        </h4>
        <img
          alt="Caption 1"
          class="Card-image"
          src="image1.jpg"
        />
        <small
          class="Card-small"
        >
          Image 1 of 3.
        </small>
      </div>
      <i
        aria-label="right arrow"
        class="bi bi-arrow-right-circle"
        role="button"
      />
    </div>
  </div>
</DocumentFragment>
`;
